packages/zwave-js/src/lib/node/Node.ts

```diff
@@ -1930,17 +1987,99 @@ protocol version:      ${this.protocolVersion}`;
-			) {
-				// Security is always supported *securely*
-				endpoint.addCC(CommandClasses.Security, { secure: true });
+			// It has been found that legacy nodes do not always advertise the S0 Command Class in their Multi
+			// Channel Capability Report and still accept all their Command Class using S0 encapsulation.
+			// A controlling node SHOULD try to control End Points with S0 encapsulation even if S0 is not
+			// listed in the Multi Channel Capability Report.
+
+			const endpointMissingS0 =
+				securityClass === SecurityClass.S0_Legacy &&
+				this.supportsCC(CommandClasses.Security) &&
+				!endpoint.supportsCC(CommandClasses.Security);
+
+			if (endpointMissingS0) {
+				this.driver.controllerLog.logNode(this.nodeId, {
+					endpoint: endpoint.index,
+					message: `is included using Security S0, but endpoint ${endpoint.index} does not list the CC. Testing if it accepts secure commands anyways.`,
+					level: "silly",
+				});
+
+				// Define which CCs we can use to test this - and if supported, how
+				const tests: {
+					ccId: CommandClasses;
+					// The test must return a truthy value if the check was successful
+					test: () => Promise<unknown>;
+				}[] = [
+					{
+						ccId: CommandClasses["Z-Wave Plus Info"],
+						test: () =>
+							endpoint.commandClasses["Z-Wave Plus Info"].get(),
+					},
+					{
+						ccId: CommandClasses["Binary Switch"],
+						test: () =>
+							endpoint.commandClasses["Binary Switch"].get(),
+					},
+					{
+						ccId: CommandClasses["Binary Sensor"],
+						test: () =>
+							endpoint.commandClasses["Binary Sensor"].get(),
+					},
+					{
+						ccId: CommandClasses["Multilevel Switch"],
+						test: () =>
+							endpoint.commandClasses["Multilevel Switch"].get(),
+					},
+					{
+						ccId: CommandClasses["Multilevel Sensor"],
+						test: () =>
+							endpoint.commandClasses["Multilevel Sensor"].get(),
+					},
+					// TODO: add other tests if necessary
+				];
+
+				for (const { ccId, test } of tests) {
+					if (!endpoint.supportsCC(ccId)) continue;
+
+					// Temporarily mark the CC as secure so we can use it to test
+					endpoint.addCC(ccId, { secure: true });
+
+					// Perform the test and treat errors as negative results
+					const success = !!(await test().catch(() => false));
+
+					if (success) {
+						this.driver.controllerLog.logNode(this.nodeId, {
+							endpoint: endpoint.index,
+							message: `Endpoint ${endpoint.index} accepts/expects secure commands`,
+							level: "silly",
+						});
+						// Mark all endpoint CCs as secure
+						for (const [ccId] of endpoint.getCCs()) {
+							endpoint.addCC(ccId, { secure: true });
+						}
+					} else {
+						this.driver.controllerLog.logNode(this.nodeId, {
+							endpoint: endpoint.index,
+							message: `Endpoint ${endpoint.index} is actually not using S0`,
+							level: "silly",
+						});
+						// Mark the CC as not secure again
+						endpoint.addCC(ccId, { secure: false });
+					}
```

Comment:
From the code above, it looks like you wanted to break here (and not continue looping for other tests).

Reply:
https://github.com/zwave-js/node-zwave-js/pull/4993
thanks :)